main.py

```diff
@@ -99,21 +101,50 @@ async def run(
     await page.goto(url)
 
     # Focus on and click the map element
-    await page.focus('.mapboxgl-canvas')
-    await page.click('.mapboxgl-canvas')
+    await asyncio.gather(page.focus('.mapboxgl-canvas'), page.click('.mapboxgl-canvas'))
+
     await asyncio.gather(
         page.evaluate(
             """
-            () => (window.performance.mark("benchmark:start"))
+            () => (window.performance.mark("benchmark-initial-load:start"))
             """
         ),
         page.click('//div[text()="Display"]/following-sibling::button'),
     )
 
     # Wait for the map to be idle and then stop timer
     await mark_and_measure(
-        page=page, start_mark='benchmark:start', end_mark='benchmark:end', label='bench'
+        page=page,
+        start_mark='benchmark-initial-load:start',
+        end_mark='benchmark-initial-load:end',
+        label='benchmark-initial-load',
     )
+
+    await asyncio.gather(page.focus('.mapboxgl-canvas'), page.click('.mapboxgl-canvas'))
+
+    if zoom_level:
+        for level in range(zoom_level):
+            start_mark = f'benchmark-{action}-level-{level}:start'
+            end_mark = f'benchmark-{action}-level-{level}:end'
+            label = f'benchmark-{action}-level-{level}'
+            if action == 'zoom_in':
+                await page.keyboard.press('=')
+                await page.evaluate(
+                    f"""
+                        () => (window.performance.mark("{start_mark}"))
+                        """
+                ),
+
+            elif action == 'zoom_out':
+                await page.keyboard.press('-'),
+                await page.evaluate(
+                    f"""
+                        () => (window.performance.mark("{start_mark}"))
+                        """
+                )
+
+            await mark_and_measure(page=page, start_mark=start_mark, end_mark=end_mark, label=label)
```

Comment:
@maxrjones / @katamartin 

i have implemented a more generic version that allows us to inject custom marks (for different actions) in the trace. this way, we can have access to the raw data for any action performed.


unfortunately, this isn't working yet. i'm running into a strange bug: i'm seeing duplicate calls to `window._map.onIdle` callback from same action/events. 

```bash
python main.py --action zoom_out --zoom-level 2
``` 

returns 

```bash
Browser console: window._map.onIdle callback called by event with label: benchmark-initial-load
Browser console: window._map.onIdle callback called by event with label: benchmark-zoom_out-level-0
Browser console: window._map.onIdle callback called by event with label: benchmark-initial-load
Browser console: window._map.onIdle callback called by event with label: benchmark-zoom_out-level-0
Browser console: window._map.onIdle callback called by event with label: benchmark-zoom_out-level-1
```

@katamartin, i'm curious... do you have any idea why this is happening? could it be that `onIdle` events from previous zoom levels are still firing while I'm trying to measure the next zoom level?

Reply:
additionally, some of the runs time out after the 5000 ms TIMEOUT. 


```bash
Browser console: window._map.onIdle callback called by event with label: benchmark-initial-load
Browser console: Max vertices per segment is 65535: bucket requested 88770
Browser console: Max vertices per segment is 65535: bucket requested 88770
Browser console: Error in page.evaluate: No idle events seen after 5000ms
1 timed out : Error in page.evaluate: No idle events seen after 5000ms
```

@katamartin, i could use your help to figure out the root cause :) Is the  `Timeout` set in the Promise below cumulative? 

```javascript
    return new Promise((resolve, reject) => {{
        const THRESHOLD = 5000
        // timeout after THRESHOLD ms if idle event is not seen
        setTimeout(() => {{
            window._error = `No idle events seen after ${{THRESHOLD}}ms`;
            reject(window._error)
        }}, THRESHOLD)
        window._map.onIdle(() => {{
            console.log(`window._map.onIdle callback called by event with label: {label}`)
            window.performance.mark('{end_mark}')
            window.performance.measure('{label}', '{start_mark}', '{end_mark}')

            resolve()
        }})
```

Reply:
> could it be that onIdle events from previous zoom levels are still firing while I'm trying to measure the next zoom level?

Aha so the `onIdle` interface is pretty crude (it's like 10 lines of code [here](https://github.com/carbonplan/maps/pull/98/files#diff-0c9683b478ef946ca030b523b405dbe4f4072e6c62edf35a6193c7d28ad1381bR5-R18)), so there's currently no way to "unregister" an idle listener, so I think what you're seeing is expected:

```
// idle after zoom 0
Browser console: window._map.onIdle callback called by event with label: benchmark-initial-load
Browser console: window._map.onIdle callback called by event with label: benchmark-zoom_out-level-0

// idle after zoom 1
Browser console: window._map.onIdle callback called by event with label: benchmark-initial-load
Browser console: window._map.onIdle callback called by event with label: benchmark-zoom_out-level-0
Browser console: window._map.onIdle callback called by event with label: benchmark-zoom_out-level-1
```
We could update the `_map` object to have a way to unregister events, but I'm torn because (1) continuing to depend on and add functionality to a specific interface might be hard to unwind for testing other mapping libraries and (2) as mentioned in that initial PR, the `onIdle` event does not _directly_ capture when rendering is completed, rather just relies on loading state tracking, which seems to functionally capture what we're interested in. This seems better suited to the "validation" + exploration stage we've been in so far, but now that we have more confidence using the chromium trace (and presumably more access to data about when the page goes static).

Reply:
> additionally, some of the runs time out after the 5000 ms TIMEOUT.

I think there might be a bug in that Promise logic. It looks like the `setTimeout` logic could ~always execute, since we're never clearing the timeout after the callback is invoked. I wonder if you'll still see the errors if you do something like:

```
    return new Promise((resolve, reject) => {{
        const THRESHOLD = 5000
        // timeout after THRESHOLD ms if idle event is not seen
        const timeoutID = setTimeout(() => {{
            window._error = `No idle events seen after ${{THRESHOLD}}ms`;
            reject(window._error)
        }}, THRESHOLD)
        window._map.onIdle(() => {{
            console.log(`window._map.onIdle callback called by event with label: {label}`)
            window.performance.mark('{end_mark}')
            window.performance.measure('{label}', '{start_mark}', '{end_mark}')
            
            clearTimeout(timeoutID) 
            resolve()
        }})
    }})
```

Reply:
I think we should migrate away from the `_map.onIdle` approach given the limitations Kata mentioned above. Please take a look at https://github.com/carbonplan/benchmark-maps/pull/24 for an alternate approach when you get a chance.